### PR TITLE
feat: defer txpool reorg until worker fetches txns for the next block

### PR DESCRIPTION
**Pull request by omerfirmak**
## 1. Purpose or design rationale of this PR

Currently we allow txpool to be reorged immediately after sequencer commits to a new block, then we end up waiting for reorg to be finished to be able to fetch new txns for the next block.

This PR forces reorg to be run in parallel with txn processing in hopes of reducing the time worker waits for it to be finished.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: core/tx_pool.go [hunk 1]
```diff
@@ -1160,13 +1161,14 @@ func (pool *TxPool) scheduleReorgLoop() {
...
 		launchNextRun bool
+		reorgsPaused  bool
 		reset         *txpoolResetRequest
 		dirtyAccounts *accountSet
 		queuedEvents  = make(map[common.Address]*txSortedMap)
 	)
 	for {
 		// Launch next background reorg if needed
-		if curDone == nil && launchNextRun {
+		if curDone == nil && (launchNextRun && !reorgsPaused) {
```
Comment (0xmountaintop) [inline core/tx_pool.go]: should we use atomic operation for the read/write?

File: core/tx_pool.go [hunk 2]
```diff
@@ -1218,6 +1220,7 @@ func (pool *TxPool) scheduleReorgLoop() {
 			}
 			close(nextDone)
 			return
+		case reorgsPaused = <-pool.reorgPauseCh:
```
Comment (0xmountaintop) [inline core/tx_pool.go:1224]: ditto

File: core/tx_pool.go (hunk 1 above)
Comment (omerfirmak) [reply core/tx_pool.go]: `scheduleReorgLoop` is a singleton, so all reads and writes to `reorgsPaused` happen within the same goroutine. So there is no concurrency going on here.

**Review by 0xmountaintop — DISMISSED** [review]

**Review by 0xmountaintop — DISMISSED** [review]

**Review by 0xmountaintop — DISMISSED** [review]

File: core/tx_pool.go [hunk 3]
```diff
@@ -1677,6 +1681,22 @@ func (pool *TxPool) demoteUnexecutables() {
 	}
 }
 
+// PauseReorgs stops any new reorg jobs to be started but doesn't interrupt any existing ones that are in flight
```
Comment (jonastheis) [inline core/tx_pool.go:1684]: Would be good to mention in the comment that these methods are blocking. For now it's not an issue but might lead to issues/unexpected behavior if multiple callers use them.

**Review by jonastheis — DISMISSED** [review]

File: core/tx_pool.go (hunk 3 above)
Comment (omerfirmak) [reply core/tx_pool.go:1684]: done

**Review by jonastheis — APPROVED** [review]

**Review by 0xmountaintop — APPROVED** [review]

**Pull request merged**